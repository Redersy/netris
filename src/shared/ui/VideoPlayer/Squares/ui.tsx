import { useEffect, useState } from "react";
import { List } from "./list";
import { SquaresType } from "../model";

export const Squares = ({ list, videoRef }: SquaresType) => {
  const [squares, setSquares] = useState<typeof list>([]);

  useEffect(() => {
    const handleProgress = () => {
      if (videoRef?.current) {
        const time = videoRef?.current?.currentTime * 1000;
        const filtered = list.filter((item) => item.start <= time && time <= item.finish);

        setSquares(filtered);
      }
    };

    videoRef?.current?.addEventListener("timeupdate", () => handleProgress());
    return videoRef?.current?.removeEventListener("timeupdate", () => handleProgress());
  }, [list, squares, videoRef]);

  return <List list={squares} />;
};
