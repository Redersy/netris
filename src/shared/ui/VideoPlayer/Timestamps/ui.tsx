import React from "react";
import { getTime, TimestampsType } from "../model";

export const Timestamps = ({ list, isLoading, error, videoRef }: TimestampsType) => {
  const handleChangeTime = (time: number) => {
    videoRef?.current && (videoRef.current.currentTime = time);
  };

  if (isLoading) return <>Loading...</>;
  if (!!error) return <>{error}</>;

  return (
    <div className="timestamps">
      {list.map((item) => {
        const { minutesWithPad, secondsWithPad, millisecondsWithPad } = getTime(item.start);
        return (
          <div
            key={`timestamp-${item.id}`}
            data-testid={`timestamp-${item.id}`}
            className="timestamps__item"
            onClick={() => handleChangeTime(item.start / 1000)}
          >{`${minutesWithPad}:${secondsWithPad}:${millisecondsWithPad}`}</div>
        );
      })}
    </div>
  );
};
