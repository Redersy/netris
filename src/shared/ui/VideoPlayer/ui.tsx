import { Timestamps } from "./Timestamps";
import { Squares } from "./Squares";
import React, { useRef } from "react";
import { VideoProps } from "./model";
import "./styles.css";

export const VideoPlayer = ({ timestamps, ...props }: VideoProps) => {
  const ref = useRef<HTMLVideoElement>(null);

  const handleVideoClick = () => {
    if (ref.current) {
      ref.current.paused ? ref.current.play() : ref.current.pause();
    }
  };

  return (
    <div className="video-player">
      <div className="video-player__wrapper" onClick={handleVideoClick}>
        <Squares list={timestamps?.list ?? []} videoRef={ref} />
        <video {...props} ref={ref} data-testid="video" />
      </div>

      <Timestamps
        list={timestamps?.list ?? []}
        isLoading={timestamps?.isLoading}
        error={timestamps?.error}
        videoRef={ref}
      />
    </div>
  );
};
