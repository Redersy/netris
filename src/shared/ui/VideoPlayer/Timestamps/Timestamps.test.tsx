import React from "react";
import { render, screen } from "@testing-library/react";
import { Timestamps } from "./ui";

const list = [
  {
    id: 1,
    zone: {
      width: 10,
      height: 10,
      left: 0,
      top: 0
    },
    start: 1,
    finish: 100
  }
];

describe("Timestamps", () => {
  test("Render", () => {
    render(<Timestamps list={list} />);

    const element = screen.getByTestId("timestamp-1");
    expect(element).toBeInTheDocument();
  });
  test("Loading", () => {
    const { asFragment } = render(<Timestamps list={list} isLoading={true} />);

    expect(asFragment()).toHaveTextContent("Loading");
  });
  test("Error", () => {
    const { asFragment } = render(<Timestamps list={list} error="Error text" />);

    expect(asFragment()).toHaveTextContent("Error text");
  });
  test("Snapshot", () => {
    const { asFragment } = render(<Timestamps list={list} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
