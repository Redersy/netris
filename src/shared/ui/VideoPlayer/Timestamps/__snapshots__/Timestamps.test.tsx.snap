// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timestamps Snapshot 1`] = `
<DocumentFragment>
  <div
    class="timestamps"
  >
    <div
      class="timestamps__item"
      data-testid="timestamp-1"
    >
      00:00:001
    </div>
  </div>
</DocumentFragment>
`;
